api/query/atoms.go

```diff
@@ -15,61 +15,118 @@ import (
 
 var browsers = shared.GetDefaultBrowserNames()
 
-type query interface {
-	toPlan(runIDs []int64) plan
+// AbstractQuery is an intermetidate representation of a test results query that
+//  has not been bound to specific shared.TestRun specs for processing.
+type AbstractQuery interface {
+	BindToRuns(runs []shared.TestRun) ConcreteQuery
 }
 
-type runQuery struct {
+// RunQuery is the internal representation of a query recieved from an HTTP
+// client, including the IDs of the test runs to query, and the structured query
+// to run.
+type RunQuery struct {
 	runIDs []int64
-	query
+	AbstractQuery
 }
 
-func (rq runQuery) toPlan() plan {
-	return rq.query.toPlan(rq.runIDs)
+// TestNamePattern is a query atom that matches test names to a pattern string.
+type TestNamePattern struct {
+	Pattern string
 }
 
-type testNamePattern struct {
-	pattern string
-}
-
-func (tnp testNamePattern) toPlan(runIDs []int64) plan {
-	return nil
+// BindToRuns for TestNamePattern is a no-op: TestNamePattern implements both
+// AbstractQuery and ConcreteQuery because it is independent of test runs.
+func (tnp TestNamePattern) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	return tnp
 }
 
 type testStatusConstraint struct {
 	browserName string
 	status      int64
 }
 
-func (tsc testStatusConstraint) toPlan(runIDs []int64) plan {
-	return nil
+func (tsc testStatusConstraint) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	ids := make([]int64, 0, len(runs))
```

Comment:
Nits:
- 1) Remind me whether there was a reason not to `make([]int64, len(runs))` and just set `ids[i] = run.ID`
- 2) (makes 1) redundant) there's a method on `shared.TestRuns` for getting the IDs. Might be called `TestRunIDs()` ?

Reply:
We do this because we only grow the slice when `run.BrowserName == tsc.browserName` (not unconditionally).